handbook/engineering/cloud/index.md

```diff
@@ -6,6 +6,26 @@ The cloud team owns all work that is necessary to build, secure, scale, and oper
 
 1. Any user or organization can use Sourcegraph Cloud with both the private and public code that they care about in a secure way.
 
+## Milestones
+
+1. Any user can add public code
+2. Organizations can use Sourcegraph with up to 50 repos or 100k LoC of private repositories for free
+    - We need to build a scalable and reliable service to count lines of code and integrate that with the rest of our systems (to enforce limits). This would align with how we’re thinking about billing by LoC, not number of repositories.
+    - No plaintext tokens or secrets anywhere (including gitserver .git/config remotes).
+    - All private repository contents need to be encrypted at rest.
+        - Gitserver.
+        - Searcher and code intel caches.
+        - Any other service that needs access to repository contents needs to be able to decrypt what it needs on the fly.
```

Comment:
Zoekt? Encrypted indexes sounds untenable. I also worry in general about how much effort this is implying and balancing that against cost of simplicity and performance. Why do we have such strict encryption goals? Is it not enough that the backing block store / FS is encrypted? What are we actually trying to achieve?

Reply:
@chayim's RFC 214 should capture the rationale for needing to encrypt repos in general, but we need to discuss what threat model we're protecting against, why, and make that explicit. Block storage encryption doesn't protect against many attack vectors (e.g. lateral movement). In your view, @keegancsmith, what are the threat models that we absolutely need to protect against, and how, and what can we get by without for now?

Reply:
Reading the proposal it sounds like we want to do this at the operating system level via encrypted filesystems. This sounds like something we could support in a somewhat generic way (although every service would need modification). However, it isn't clear to me the performance impact and which specific cases this protects us vs relying on GCP disk encryption + better hygiene around service permissions.

To me this sounds like it is protecting against a bad actor getting hold of our GCP account in such a way where they can't get into our cluster, but they can maybe decrypt disk backups/mount disks? If they can get into the cluster, one can only assume that if the service can read it, the attacker can read it. How much of a threat is that really? What does GCP provide to protect against that? (I assume locking down IAM will help a lot here).

Just as an aside we have lots of service which rely on mmap a blob of disk which contains repository file contents. This implies that the encryption layer will need to be at the OS level, otherwise introducing it will be far too much effort. The alternative that jumps into my head is properly authenticating service actors, ensuring services which persist data also respect service actors, and for sensitive data sharding data in a privacy sensitive manor.

Reply:
@chayim FYI I hadn't seen RFC 214 before. I think you need to add a comment with someones email for them to be notified. The RFC looks really good, my only point of discussion is number 6 about encrypting repo contents. Don't want to derail the rest of whats a really great plan.

Reply:
> Zoekt? Encrypted indexes sounds untenable. I also worry in general about how much effort this is implying and balancing that against cost of simplicity and performance. Why do we have such strict encryption goals? Is it not enough that the backing block store / FS is encrypted? What are we actually trying to achieve?

Encrypting indexes is very hard - unless the indexes are in memory. Scanning encrypted in memory data is at best a pain, and starts to become that tradeoff of nice to have security versus a need. I'd argue that specific customers might **need** it, but generally it's a nice to have - similar to your stance. For my own education and thinking, @keegancsmith does Zoekt store indices in memory, or flush them back to disk?

> @chayim FYI I hadn't seen RFC 214 before. I think you need to add a comment with someones email for them to be notified. The RFC looks really good, my only point of discussion is number 6 about encrypting repo contents. Don't want to derail the rest of whats a really great plan.

Thanks for the vote of confidence - especially when it's incomplete!  We were just discussing email notification of key rollover with regards to RFC196, and this might be an opportunity to kill two birds with one service.  I know FUSE is something that appeals to you and I wanted to drag you in @keegancsmith .

> @chayim's [RFC 214](https://docs.google.com/document/d/1ajVQ11Vyi0Wz67DN15W3girJ5kyB5Zb7vk0LK6r7gbQ/edit) should capture the rationale for needing to encrypt repos in general, but we need to discuss what threat model we're protecting against, why, and make that explicit. Block storage encryption doesn't protect against many attack vectors (e.g. lateral movement). In your view, @keegancsmith, what are the threat models that we absolutely need to protect against, and how, and what can we get by without for now?

Absolutely agreed. This is the beginning of the RFC for that reason - and thank you for affirming that need.  I'll make it a point to get that into the background information area, and explicitly delegate a threat model.